validate/validator.py

```diff
@@ -0,0 +1,612 @@
+                    return num_template(option_value, field.name, option_value.sint32)
+                elif str(option_value.sint64):
+                    return num_template(option_value, field.name, option_value.sint64)
+                elif str(option_value.fixed32):
+                    return num_template(option_value, field.name, option_value.fixed32)
+                elif str(option_value.fixed64):
+                    return num_template(option_value, field.name, option_value.fixed64)
+                elif str(option_value.sfixed32):
+                    return num_template(option_value, field.name, option_value.sfixed32)
+                elif str(option_value.sfixed64):
+                    return num_template(option_value, field.name, option_value.sfixed64)
+                else:
+                    return "raise UnimplementedException()"
+    if field.message_type and not field.containing_oneof:
+        for option_descriptor, option_value in field.GetOptions().ListFields():
+            if option_descriptor.full_name == "validate.rules":
+                if str(option_value.duration):
+                    return duration_template(option_value, field.name)
+                elif str(option_value.timestamp):
+                    return timestamp_template(option_value, field.name)
+                elif str(option_value.float) or str(option_value.int32) or str(option_value.int64) or \
+                        str(option_value.double) or str(option_value.uint32) or str(option_value.uint64) or \
+                        str(option_value.bool) or str(option_value.string):
+                    return wrapper_template(option_value, field.name)
+                elif str(option_value.bytes):
+                    return "raise UnimplementedException()"
+                elif str(option_value.message) is not "":
+                    return message_template(option_value, field.name)
+                else:
+                    return "raise UnimplementedException()"
+        if field.message_type.full_name.startswith("google.protobuf"):
+            return ""
+        else:
+            return message_template(None, field.name)
+    return ""
+
+def file_template(proto_message):
+    file_tmp = """
+# Validates {{ p.DESCRIPTOR.name }}
+def validate(p):
+    {%- for option_descriptor, option_value in p.DESCRIPTOR.GetOptions().ListFields() %}
+        {%- if option_descriptor.full_name == "validate.disabled" and option_value %}
+    return None
+        {%- endif -%}
+    {%- endfor -%}
+    {%- for field in p.DESCRIPTOR.fields -%}
+        {%- if field.label == 3 or field.containing_oneof %}
+    raise UnimplementedException()
+        {%- else %}
+    {{ rule_type(field) -}}
+        {%- endif -%}
+    {%- endfor %}
+    return None"""
+    return Template(file_tmp).render(rule_type = rule_type, p = proto_message)
+
+class UnimplementedException(Exception):
+    pass
+
+class ValidationFailed(Exception):
```

Comment:
Please take a look at the exceptions from the other languages and include the other ancillary information that will help developers track down the source of their errors.

Reply:
Have included the ancillary info while raising the exception. For example here:
https://github.com/envoyproxy/protoc-gen-validate/pull/238/files#diff-2c27ae30e4a405617a8988599f7e85f9R88
```
raise ValidationFailed(\"{{ name }} not equal to {{ o.string['const'] }}\")
```
This can then be caught and read as follows:
```
except ValidationFailed as e:
    reason = str(e)
```
Please let me know if you think I missed anything here.